### ISSUE TITLE
KissFFT

### ISSUE THREAD

**Issue by iff** [xdata-skylark/libskylark#19]
@haimav also proposed to support kissFFT as a default FFT implementation:

http://sourceforge.net/projects/kissfft/

``` cpp
#include "kiss_fft.h"

kiss_fft_cfg cfg = kiss_fft_alloc( nfft ,is_inverse_fft ,0,0 );

// put kth sample in cx_in[k].r and cx_in[k].i

kiss_fft( cfg , cx_in , cx_out );

// transformed. DC is in cx_out[0].r and cx_out[0].i 

free(cfg);
```

**Comment by jomsdev** [comment]
I have been working on this for some time. Right now there is a version with libskylark compiling against kissfft as default (but it is "easy" change it if you want to compile using fftw). I moved the logic implemented in FUT.hpp to the folder utility/fft with three different implementations: using kissfft, using fftw, without any fft implementation (empty object that throws error if you use it).

kissfft does not have any fft function from real to real. I am compiling using the .hh file (only the header is necessary). I am trying to implement the DCT type II using 4N FFT following the next steps:

- I start with a vector [a, b, c, d] of doubles/floats
- I create a vector of complex [0, a, 0, b, 0,  c,  0,  d,  0,  d,  0,  c, 0, b, 0, a]
- I apply kissfft.transform() to the compelx vector
- After the transform the vector looks like [A, B, C, D, 0, -D, -C, -B, -A, -B, -C, -D, 0, D, C, B]
- The solution is the real part of [A,B,C,D]


However, the results are not being what I expected. After apply kissfft.transform()

The vector [-0.729046 0.670017 0.937736 -0.557932] is transformed to

 (WIP, I will edit this comment after/during the weekly call)

**Comment by gidiko** [comment]
### Things come out OK

**Code**

```
%%cython --cplus --include /vagrant/notebooks/ --annotate

from libcpp.complex cimport complex
from libcpp.vector cimport vector

import numpy as np
from scipy.fftpack import dct

# utilities for doing the expansions x4 as described in
# http://dsp.stackexchange.com/questions/2807/fast-cosine-transform-via-fft
def insert_zeros(x):
    n = len(x)
    data = [0 for i in range(2 * n)]
    for i in range(n):
        data[2 * i + 1] = x[i]
    return data

def expand4(x):
    return insert_zeros(x) + insert_zeros(x[::-1])   


cdef extern from "<kissfft.hh>":
    cdef cppclass kissfft[scalar_type]:
        kissfft(size_t, bool)
        void transform(complex[scalar_type] *, complex[scalar_type] *)

ctypedef complex[double] complex_number
   
cdef enum:
    N4 = 40

cdef kissfft[double] * fft         = new kissfft[double](N4, True)
cdef complex_number[N4] source
cdef complex_number[N4] destination

n = N4 / 4
input_data = np.random.random(n)
numbers = expand4(input_data)


for i, number in enumerate(numbers):
    source[i]  =  complex_number(number)

fft.transform(source, destination)

print '=' * 20
print '[DEBUG]: What the output x4 buffer contains'
print '-' * 20
for i in range(N4):
    print [destination[i].real(), destination[i].imag()]
print

print '=' * 20
print 'Input data'
print '-' * 20
for i in range(n):
    print input_data[i]
print

print '=' * 20
print 'DCT transformed data based on KissFFT' 
print '-' * 20
kissfft_transformed_data = [destination[i].real() for i in range(n)]
for i in range(n):
    print kissfft_transformed_data[i]


print '=' * 20
print 'DCT transformed data based on scipy.fftpack.dct' 
print '-' * 20
fftpack_transformed_data = dct(input_data)
for i in range(n):
    print fftpack_transformed_data[i]
```

**Output from running it**
====================
[DEBUG]: What the output x4 buffer contains
--------------------
[8.92475447881435, 0.0]
[-1.6339659284048371, 5.689893001203927e-16]
[-1.9818146430166106, 3.3306690738754696e-16]
[-1.5878066429230464, 4.440892098500626e-16]
[2.1770848390845394, -2.220446049250313e-16]
[0.8631617121724362, -5.551115123125783e-17]
[0.6436832901201757, 0.0]
[1.050314532983049, 0.0]
[-1.7624810045259016, 5.551115123125783e-16]
[-0.09522066656647431, -1.1102230246251565e-16]
[0.0, 0.0]
[0.09522066656647431, 0.0]
[1.762481004525902, 1.1102230246251565e-16]
[-1.050314532983049, 1.1102230246251565e-16]
[-0.6436832901201758, 0.0]
[-0.8631617121724362, 5.551115123125783e-17]
[-2.1770848390845394, 2.220446049250313e-16]
[1.5878066429230464, -1.1102230246251565e-16]
[1.981814643016611, 1.1102230246251565e-16]
[1.633965928404837, 1.249000902703301e-16]
[-8.92475447881435, 0.0]
[1.6339659284048371, -5.689893001203927e-16]
[1.9818146430166106, -3.3306690738754696e-16]
[1.5878066429230464, -4.440892098500626e-16]
[-2.1770848390845394, 2.220446049250313e-16]
[-0.8631617121724362, 5.551115123125783e-17]
[-0.6436832901201757, 0.0]
[-1.050314532983049, 0.0]
[1.7624810045259016, -5.551115123125783e-16]
[0.09522066656647431, 1.1102230246251565e-16]
[0.0, 0.0]
[-0.09522066656647431, 0.0]
[-1.762481004525902, -1.1102230246251565e-16]
[1.050314532983049, -1.1102230246251565e-16]
[0.6436832901201758, 0.0]
[0.8631617121724362, -5.551115123125783e-17]
[2.1770848390845394, -2.220446049250313e-16]
[-1.5878066429230464, 1.1102230246251565e-16]
[-1.981814643016611, -1.1102230246251565e-16]
[-1.633965928404837, -1.249000902703301e-16]

====================
Input data
--------------------
0.221625845572
0.0129673878011
0.0555177701693
0.867850694542
0.723121207687
0.713980687316
0.530661049118
0.0490445089901
0.674720112882
0.61288797533

====================
DCT transformed data based on KissFFT
--------------------
8.92475447881
-1.6339659284
-1.98181464302
-1.58780664292
2.17708483908
0.863161712172
0.64368329012
1.05031453298
-1.76248100453
-0.0952206665665

====================
DCT transformed data based on scipy.fftpack.dct
--------------------
8.92475447881
-1.6339659284
-1.98181464302
-1.58780664292
2.17708483908
0.863161712172
0.64368329012
1.05031453298
-1.76248100453
-0.0952206665665

**Comment by gidiko** [comment]
### Similarly for the comparison of FFTW's DCT to fftpack's DCT: THINGS AGREE HERE AS WELL

**The code**

```
%%cython --cplus

import numpy as np
from scipy.fftpack import dct

cdef enum:
    n = 10

cdef double[n] temp
cdef double[n] fftw_transformed_data

np.random.seed(seed=0)
input_data = np.random.random(n)

for i in range(n):
    fftw_transformed_data[i] = input_data[i]

cdef extern from "<fftw3.h>":
    cdef enum fftw_r2r_kind:
        FFTW_R2HC=0
        FFTW_HC2R=1
        FFTW_DHT=2
        FFTW_REDFT00=3
        FFTW_REDFT01=4
        FFTW_REDFT10=5
        FFTW_REDFT11=6
        FFTW_RODFT00=7
        FFTW_RODFT01=8
        FFTW_RODFT10=9
        FFTW_RODFT11=10
        
    cdef int FFTW_UNALIGNED
    cdef int FFTW_ESTIMATE
    
    ctypedef struct fftw_plan_struct:
        pass
    ctypedef fftw_plan_struct *fftw_plan

    fftw_plan fftw_plan_r2r_1d(int, double*, double*, fftw_r2r_kind, unsigned)
    void fftw_execute_r2r(fftw_plan, double*, double*)
    void fftw_destroy_plan(fftw_plan)

cdef fftw_plan plan = fftw_plan_r2r_1d(n, temp, temp, FFTW_REDFT10, FFTW_UNALIGNED | FFTW_ESTIMATE)
fftw_execute_r2r(plan, fftw_transformed_data, fftw_transformed_data)
fftw_destroy_plan(plan)

print '=' * 20
print 'Input data'
print '-' * 20
for i in range(n):
    print input_data[i]
print

print '=' * 20
print 'DCT transformed data based on fftw' 
print '-' * 20
for i in range(n):
    print fftw_transformed_data[i]
print

print '=' * 20
print 'DCT transformed data based on scipy.fftpack.dct' 
print '-' * 20
fftpack_transformed_data = dct(input_data)
for i in range(n):
    print fftpack_transformed_data[i]
```

**The output**
====================
Input data
--------------------
0.548813503927
0.715189366372
0.602763376072
0.544883182997
0.423654799339
0.645894113067
0.437587211263
0.891773000782
0.963662760501
0.383441518826

====================
DCT transformed data based on fftw
--------------------
12.3153256663
-0.496941270994
0.55749732799
0.615515874887
-1.39487372089
0.831432645803
-1.4859954164
0.661864770858
-0.07985466977
-0.761581633783

====================
DCT transformed data based on scipy.fftpack.dct
--------------------
12.3153256663
-0.496941270994
0.55749732799
0.615515874887
-1.39487372089
0.831432645803
-1.4859954164
0.661864770858
-0.07985466977
-0.761581633783

**Comment by jomsdev** [comment]
I have finished the integration of DCT using KissFFT. All the cmake files have been updated. The 
standard configuration is  -DUSEKISSFFT=ON

I have not implemented IDCT (yet) which is used in the [RFUT_t file](https://github.com/xdata-skylark/libskylark/blob/fftw/sketch/RFUT_Elemental.hpp).

Testing the code with the MNIST example the results seems to agree with the FFTW version. However the times are (as we expected) worst using the kissfft version.

- Compiled with KissFFT and  using FJLT to reduce the number of rows from 60k to 10k takes **37.5s** 

CC=mpicc CXX=mpicxx cmake -DBUILD_ML=OFF -DUSE_FFTW=OFF -DUSE_KISSFFT=ON -DUSE_COMBBLAS=ON -DBOOST_ROOT=/usr/local/include -DBUILD_EXAMPLES=OFF -DCMAKE_INSTALL_PREFIX=${SKYLARK_INSTALL_DIR} ${SKYLARK_SRC_DIR} 

- Compiled with FFTW and  using FJLT to reduce the number of rows from 60k to 10k takes **24.9s** 

CC=mpicc CXX=mpicxx cmake -DBUILD_ML=OFF -DUSE_FFTW=ON -DUSE_KISSFFT=OFF -DUSE_COMBBLAS=ON -DBOOST_ROOT=/usr/local/include -DBUILD_EXAMPLES=OFF -DCMAKE_INSTALL_PREFIX=${SKYLARK_INSTALL_DIR} ${SKYLARK_SRC_DIR}


The code used for the comparation has been:
```python
%%time

for i in range (10):
    T = 10000

    # We creat a sketch object
    S = sl_sketch.FJLT(A.Height(), T)

    # The operator * is overloaded to apply the sketch correctly
    SA = S * A # <-------- Apply on the columns

    # We are reducing the number of rows, so we need to do the same with 
    # our labels
    Seb = S * eb
```

**Comment by gidiko** [comment]
Nice. Also timings seem within the typical FFTW - KISSFFT diffs in [fftw benchmarks]( http://www.fftw.org/speed/)

**Comment by jomsdev** [comment]
DCT II implemented #53 

Waiting for the code review and merge

**Comment by jomsdev** [comment]
Reviewed and accepted.
